src/core/utils/components/VirtualScroll.tsx

```diff
@@ -100,145 +132,224 @@ export type VirtualScrollProps = {
       !scrollContainer.current ||
       scrollContainer.current === parentRef.current?.ownerDocument.body
-    ) {
-      parentRef.current?.ownerDocument.addEventListener(
-        'scroll',
-        updateVirtualScroll,
-      );
-    } else {
-      scrollContainer.current.addEventListener('scroll', updateVirtualScroll);
-    }
-    return removeScrollListener;
-  }, [updateVirtualScroll, removeScrollListener]);
-
-  React.useLayoutEffect(() => {
-    updateVirtualScroll();
-  }, [scrollContainerHeight, itemsLength, updateVirtualScroll]);
-
-  return (
-    <div
-      style={{
-        overflow: 'hidden',
-        minHeight: itemsLength * childHeight.current,
-        width: '100%',
-        ...style,
-      }}
-      ref={ref}
-      {...rest}
-    >
+        ? parentRef.current?.ownerDocument.removeEventListener(
+            'scroll',
+            onScrollRef.current,
+          )
+        : scrollContainer.current.removeEventListener(
+            'scroll',
+            onScrollRef.current,
+          );
+    }, []);
+
+    // Add event listener to the scrollable container.
+    React.useLayoutEffect(() => {
+      removeScrollListener();
+      onScrollRef.current = onScroll;
+      if (
+        !scrollContainer.current ||
+        scrollContainer.current === parentRef.current?.ownerDocument.body
+      ) {
+        parentRef.current?.ownerDocument.addEventListener('scroll', onScroll);
+      } else {
+        scrollContainer.current.addEventListener('scroll', onScroll);
+      }
+      return removeScrollListener;
+    }, [onScroll, removeScrollListener]);
+
+    React.useLayoutEffect(() => {
+      if (!scrollContainerHeight) {
+        return;
+      }
+
+      const scrollableContainer =
+        scrollContainer.current ??
+        (parentRef.current?.ownerDocument.scrollingElement as HTMLElement);
+
+      // if `scrollToIndex` is not visible, scroll to it
+      if (
+        scrollableContainer &&
+        (scrollToIndex > visibleIndex.current.end ||
+          scrollToIndex < visibleIndex.current.start)
+      ) {
+        const indexDiff =
+          scrollToIndex > visibleIndex.current.end
+            ? scrollToIndex - visibleIndex.current.end
+            : scrollToIndex - visibleIndex.current.start;
+
+        // smallest scroll to the index
+        const scrollTop =
+          scrollableContainer.scrollTop + indexDiff * childHeight.current.child;
+
+        scrollableContainer.scrollTo({
+          top:
+            Math.floor(scrollTop / childHeight.current.child) *
+            childHeight.current.child,
+        });
+      }
+    }, [scrollToIndex, scrollContainerHeight]);
+
+    React.useLayoutEffect(() => {
+      if (!scrollContainerHeight) {
+        return;
+      }
+
+      updateVirtualScroll();
+    }, [scrollContainerHeight, updateVirtualScroll]);
+
+    return (
       <div
         style={{
-          willChange: 'transform',
+          overflow: 'hidden',
+          minHeight:
+            itemsLength > 1
```

Comment:
Is this check needed? If there are no items, length === 0, that means 0 * height is 0.

Reply:
See https://github.com/iTwin/iTwinUI-react/pull/539#discussion_r800798768